File: leetcode/994.rotting-oranges.py
# @leet start
class Solution:
    def orangesRotting(self, grid: list[list[int]]) -> int:
        FRESH = 1
        ROTTING = 2

        rows = len(grid)
        cols = len(grid[0])

        level: list[tuple[int, int]] = []
        fresh_count = 0

        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == ROTTING:
                    level.append((row, col))
                elif grid[row][col] == FRESH:
                    fresh_count += 1

        # Multi-source BFS
        time = 0

        while level:
            next_level: list[tuple[int, int]] = []

            for row, col in level:
                for delta_row, delta_col in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                    adj_row = row + delta_row
                    adj_col = col + delta_col

                    if (
                        0 <= adj_row < rows
                        and 0 <= adj_col < cols
                        and grid[adj_row][adj_col] == FRESH
                    ):
                        next_level.append((adj_row, adj_col))
                        grid[adj_row][adj_col] = ROTTING
                        fresh_count -= 1

            level = next_level

            if level:
                time += 1

        return time if fresh_count == 0 else -1


# @leet end

File: leetcode/994.rotting-oranges.cpp
// @leet start
#include <algorithm>
#include <cstddef>
#include <queue>
#include <tuple>
#include <vector>

class Solution {
 public:
  int orangesRotting(std::vector<std::vector<int>>& grid) {
    std::vector<std::tuple<int, int>> dirs{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    std::queue<std::tuple<std::size_t, std::size_t, int>> q;

    for (std::size_t row = 0; row < grid.size(); row++) {
      for (std::size_t col = 0; col < grid[0].size(); col++) {
        if (grid[row][col] == 2) {
          q.push({row, col, 0});
        }
      }
    }

    int maxTime = 0;

    while (!q.empty()) {
      const auto [row, col, time] = q.front();
      q.pop();

      for (const auto [dx, dy] : dirs) {
        const int newRow = row + dx;
        const int newCol = col + dy;

        if (newRow >= 0 && newRow < grid.size() && newCol >= 0 && newCol < grid[0].size() &&
            grid[newRow][newCol] == 1) {
          grid[newRow][newCol] = 2;

          const int newTime = time + 1;
          maxTime = std::max(maxTime, newTime);
          q.push({newRow, newCol, newTime});
        }
      }
    }

    for (std::size_t row = 0; row < grid.size(); row++) {
      for (std::size_t col = 0; col < grid[0].size(); col++) {
        if (grid[row][col] == 1) {
          return -1;
        }
      }
    }

    return maxTime;
  }
};
// @leet end
